crates/aptos/src/node/fetch_metadata.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use super::analyze_validators::NewBlockInfo;
+use anyhow::Result;
+use aptos_rest_client::{
+    aptos_api_types::{BlockMetadataTransaction, IdentifierWrapper, MoveResource, WriteSetChange},
+    Client as RestClient, Transaction,
+};
+use aptos_types::account_address::AccountAddress;
+use aptos_types::account_config::NewBlockEvent;
+use std::convert::TryFrom;
+use std::str::FromStr;
+
+pub struct EpochInfo {
+    pub epoch: u64,
+    pub blocks: Vec<NewBlockInfo>,
+    pub validators: Vec<AccountAddress>,
+}
+
+pub struct FetchMetadata {}
+
+impl FetchMetadata {
+    fn metadata_to_new_block_event(metadata: &BlockMetadataTransaction) -> NewBlockInfo {
+        NewBlockInfo {
+            event: NewBlockEvent::new(
+                *metadata.epoch.inner(),
+                *metadata.round.inner(),
+                0,
+                metadata.previous_block_votes.clone(),
+                *metadata.proposer.inner(),
+                metadata
+                    .failed_proposer_indices
+                    .iter()
+                    .map(|i| *i as u64)
+                    .collect(),
+                *metadata.timestamp.inner(),
+            ),
+            version: metadata.info.version.into(),
+        }
+    }
+
+    async fn get_epoch_at_version(
+        mut cursor: u64,
+        batch: u64,
+        last_version: u64,
+        client: &RestClient,
+    ) -> Result<u64> {
+        loop {
+            let wanted = client
+                .get_transactions(Some(cursor), Some(u16::try_from(batch).unwrap()))
+                .await?
+                .into_inner()
+                .into_iter()
+                .filter(|t| matches!(t, Transaction::BlockMetadataTransaction(_)))
+                .next();
+            if let Some(t) = wanted {
+                if let Transaction::BlockMetadataTransaction(m) = t {
+                    return Ok(*m.epoch.inner());
+                }
+            }
+            cursor += batch;
+            if cursor > last_version {
+                return Ok(u64::MAX);
+            }
+        }
+    }
+
+    fn get_validator_addresses(
+        data: &MoveResource,
+        field_name: &str,
+    ) -> Option<Vec<AccountAddress>> {
+        let active_validators_json = data
+            .data
+            .0
+            .get(&IdentifierWrapper::from_str(field_name).unwrap())
+            .unwrap();
+        if let serde_json::Value::Array(active_validators) = active_validators_json {
+            let mut validators: Vec<AccountAddress> = vec![];
+            for validator in active_validators {
```

Comment:
nit: many nested if-else making this code a bit unreadable. Possible to simplify or may be create a helper function to extract the address?